### PR TITLE
rework upsert flow to handle all errors from API

### LINKED ISSUE #159: upsert() function does not correctly handle errors
using the upsert() function I got this error.

The code in the client checks if there was an existing document error, but does not check if any other error occured. 

See https://github.com/datastax/astrapy/blob/master/astrapy/db.py#L812

```
2024-01-09 07:47:04.364 - DEBUG   - astrapy.utils - unknown_worker - Response content: {"status":{"insertedIds":[]},"errors":[{"message":"Failed to insert document with _id 'recent_articles': Cassandra timeout during CAS write query at consistency LOCA
L_SERIAL (2 replica were required but only 1 acknowledged the write)"}]}
2024-01-09 07:47:04.364 - ERROR   - root - store_article_diff-1 - Error in worker store_article_diff-1
Traceback (most recent call last):
  File "/Users/aaron.morton/code/riptano/wikipedia-scripts/v2/utils.py", line 52, in _worker
    result = await self.func(item)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaron.morton/code/riptano/wikipedia-scripts/v2/process.py", line 266, in store_article_diff
    await update_article_metadata(article_diff)
  File "/Users/aaron.morton/code/riptano/wikipedia-scripts/v2/process.py", line 373, in update_article_metadata
    resp = await utils.wrap_blocking_io(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaron.morton/code/riptano/wikipedia-scripts/v2/utils.py", line 19, in wrap_blocking_io
    return await loop.run_in_executor(None, func, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaron.morton/.pyenv/versions/3.12.1/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaron.morton/code/riptano/wikipedia-scripts/v2/process.py", line 374, in <lambda>
    lambda x: SUGGESTIONS_COLLECTION.upsert(x),
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaron.morton/code/riptano/wikipedia-scripts/.venv/lib/python3.12/site-packages/astrapy/db.py", line 720, in upsert
    upserted_id = cast(str, result["status"]["insertedIds"][0])
                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range

```

### PR DESCRIPTION
**Pull request by hemidactylus**
Fixes #159  .

The errors in CI (apart from timeouts and such) are due to the "ordered=false" default - unrelated to this PR - #169 takes care of them.

### CODE REVIEW

**Comment by hemidactylus** [comment]
@amorton FYI, and thank you for surfacing this missing error handling in the flow!

File: tests/astrapy/test_async_db_dml.py [hunk 1]
```diff
@@ -650,6 +650,40 @@ async def test_upsert_document(
...
+    assert upsert_result0 == _id0
+
+    # triggering an API error for the already-exists path of the upsert
+    document0b = {
+        "_id": _id0,
+        "nature": "faulty vector",
+        "$vector": [10, 11, 999, -153],
+    }
+    with pytest.raises(ValueError):
+        _ = await async_writable_vector_collection.upsert(document0b)
```
Comment (cbornet) [inline tests/astrapy/test_async_db_dml.py:675]:
```suggestion
        await async_writable_vector_collection.upsert(document0b)
```

File: tests/astrapy/test_async_db_dml.py [hunk 2]
```diff
@@ -650,6 +650,40 @@ async def test_upsert_document(
...
+
+    # triggering an API error for the already-exists path of the upsert
+    document1 = {
+        "_id": _id1,
+        "nature": "faulty vector from the start",
+        "$vector": [10, 11, 999, -153],
+    }
+    with pytest.raises(ValueError):
+        _ = await async_writable_vector_collection.upsert(document1)
```
Comment (cbornet) [inline tests/astrapy/test_async_db_dml.py:684]:
```suggestion
        await async_writable_vector_collection.upsert(document1)
```

File: tests/astrapy/test_db_dml.py [hunk 3]
```diff
@@ -698,6 +698,38 @@ def test_upsert_document(writable_vector_collection: AstraDBCollection) -> None:
...
+    assert upsert_result0 == _id0
+
+    # triggering an API error for the already-exists path of the upsert
+    document0b = {
+        "_id": _id0,
+        "nature": "faulty vector",
+        "$vector": [10, 11, 999, -153],
+    }
+    with pytest.raises(ValueError):
+        _ = writable_vector_collection.upsert(document0b)
```
Comment (cbornet) [inline tests/astrapy/test_db_dml.py:721]:
```suggestion
        writable_vector_collection.upsert(document0b)
```

File: tests/astrapy/test_db_dml.py [hunk 4]
```diff
@@ -698,6 +698,38 @@ def test_upsert_document(writable_vector_collection: AstraDBCollection) -> None:
...
+
+    # triggering an API error for the already-exists path of the upsert
+    document1 = {
+        "_id": _id1,
+        "nature": "faulty vector from the start",
+        "$vector": [10, 11, 999, -153],
+    }
+    with pytest.raises(ValueError):
+        _ = writable_vector_collection.upsert(document1)
```
Comment (cbornet) [inline tests/astrapy/test_db_dml.py:730]:
```suggestion
        writable_vector_collection.upsert(document1)
```

**Review by erichare — APPROVED** [review]

**Pull request merged**